src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,157 @@
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.cbegin(), choices.cend());
+
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
+  });
+
+  // If current state exceeds cost_budget,
+  // reject the least desirable choices to reduce cost_balance.
+  if (cost_balance > cost_budget) {
+    LOG_INFO("Cost balance of " << cost_balance << " exceeds budget of " << cost_budget);
+    const auto sacrifice_until = sacrifice_choices(sorted_choices, cost_budget - cost_balance);
+    if (sacrifice_until == sorted_choices.cend()) {
+      LOG_WARN("Cost budget is impossible to maintain. No Operations are performed.");
+      return operations;
+    }
+    for (auto choice = sorted_choices.cbegin(); choice != sacrifice_until; ++choice) {
+      LOG_DEBUG(" ! Reject " << **choice << " to reduce cost balance");
+      operations.push_back((*choice)->reject());
+      cost_balance += (*choice)->reject_cost();
+      desirability_balance += (*choice)->reject_desirability();
+    }
+    sorted_choices.erase(sorted_choices.cbegin(), sacrifice_until);
+  }
+
+  // Select the most desirable operations first
+  // always maintaining the cost budget
+  while (sorted_choices.size() > 0) {
+    if (sorted_choices.front()->reject_desirability() > sorted_choices.back()->accept_desirability()) {
+      LOG_DEBUG("Rejecting " << *sorted_choices.front() << " is most beneficial.");
+
+      /*
+       * Rejecting a choice can only reduce cost_balance (i.e. free up resources)
+       * and never exceed cost_budget (no new costs will be added) as cost() >= 0:
+       * Rejecting a choice that already exists yields ("gives back") the cost
+       * that was paid for it.
+       * Rejecting a choice that does not yet exist cannot cost anything since
+       * nothing will be changed.
+       */
+
+      LOG_DEBUG(" ! Reject " << *sorted_choices.front());
+      operations.push_back(sorted_choices.front()->reject());
+      cost_balance += sorted_choices.front()->reject_cost();
+      desirability_balance += sorted_choices.front()->reject_desirability();
+      sorted_choices.pop_front();
+
+    } else {
+      LOG_DEBUG("Accepting " << *sorted_choices.back() << " is most beneficial.");
+
+      /*
+       * Accepting a choice could exceed cost_budget (cost more resources than
+       * there are available), so try to reduce cost_balance first (free up used resources)
+       */
+
+      const auto sacrifice_until =
+          sacrifice_choices(sorted_choices, cost_budget - cost_balance - sorted_choices.back()->accept_cost(),
+                            -sorted_choices.back()->accept_desirability());
```

Comment:
Lost again. Didn‘t you do the sacrifices already? I don‘t get what happens in this whole block - maybe it gets clearer when we have improved the naming.